get-started/09_image_best.md

```diff
@@ -0,0 +1,274 @@
+install the dependencies, and _then_ copy in everything else? Then, we only recreate the yarn dependencies if there was
+a change to the `package.json`. Make sense?
+
+1. Update the Dockerfile to copy in the `package.json` first, install dependencies, and then copy everything else in.
+
+    ```dockerfile hl_lines="3 4 5"
+    FROM node:12-alpine
+    WORKDIR /app
+    COPY package.json yarn.lock ./
+    RUN yarn install --production
+    COPY . .
+    CMD ["node", "src/index.js"]
+    ```
+
+1. Create a file named `.dockerignore` in the same folder as the Dockerfile with the following contents.
+
+    ```ignore
+    node_modules
+    ```
+
+    `.dockerignore` files are an easy way to selectively copy only image relevant files.
+    You can read more about this
+    [here](https://docs.docker.com/engine/reference/builder/#dockerignore-file).
+    In this case, the `node_modules` folder should be omitted in the second `COPY` step because otherwise,
+    it would possibly overwrite files which were created by the command in the `RUN` step.
+    For further details on why this is recommended for Node.js applications and other best practices,
+    have a look at their guide on
+    [Dockerizing a Node.js web app](https://nodejs.org/en/docs/guides/nodejs-docker-webapp/).
+
+1. Build a new image using `docker build`.
+
+    ```bash
+    docker build -t getting-started .
+    ```
+
+    You should see output like this...
+
+    ```plaintext
+    Sending build context to Docker daemon  219.1kB
+    Step 1/6 : FROM node:12-alpine
+    ---> b0dc3a5e5e9e
+    Step 2/6 : WORKDIR /app
+    ---> Using cache
+    ---> 9577ae713121
+    Step 3/6 : COPY package.json yarn.lock ./
+    ---> bd5306f49fc8
+    Step 4/6 : RUN yarn install --production
+    ---> Running in d53a06c9e4c2
+    yarn install v1.17.3
+    [1/4] Resolving packages...
+    [2/4] Fetching packages...
+    info fsevents@1.2.9: The platform "linux" is incompatible with this module.
+    info "fsevents@1.2.9" is an optional dependency and failed compatibility check. Excluding it from installation.
+    [3/4] Linking dependencies...
+    [4/4] Building fresh packages...
+    Done in 10.89s.
+    Removing intermediate container d53a06c9e4c2
+    ---> 4e68fbc2d704
+    Step 5/6 : COPY . .
+    ---> a239a11f68d8
+    Step 6/6 : CMD ["node", "src/index.js"]
+    ---> Running in 49999f68df8f
+    Removing intermediate container 49999f68df8f
+    ---> e709c03bc597
+    Successfully built e709c03bc597
+    Successfully tagged getting-started:latest
+    ```
+
+    You'll see that all layers were rebuilt. Perfectly fine since we changed the Dockerfile quite a bit.
+
+1. Now, make a change to the `src/static/index.html` file (like change the `<title>` to say "The Awesome Todo App").
+
+1. Build the Docker image now using `docker build -t getting-started .` again. This time, your output should look a little different.
+
+    ```plaintext hl_lines="5 8 11"
```

Comment:
needs formatting.  It appears as a single line in the preview

Reply:
fixed